alphacsc/online_dictionary_learning.py

```diff
@@ -62,52 +63,46 @@ def partial_fit(self, X, y=None):
         self._check_param_partial_fit()
         self._ensure_fit_init(X)
 
-        # Compute the activations for the current batch and get the sufficient
-        # statistic for the dictionary update
-        z_hat, ztz, ztX = update_z_multi(
-            X, self._D_hat, reg=self.reg_,
-            solver=self.solver_z, solver_kwargs=self.solver_z_kwargs,
-            loss=self.loss, loss_params=self.loss_params,
-            n_jobs=self.n_jobs, return_ztz=True)
+        with get_z_encoder_for(X, self._D_hat, self.n_atoms, self.n_times_atom,
+                               self.n_jobs, self.solver_z,
+                               self.solver_z_kwargs, self.reg_, self.loss,
+                               self.loss_params) as z_encoder:
 
-        alpha = self.algorithm_params['alpha']
-        self.constants['XtX'] = X.ravel().dot(X.ravel())
-        self.constants['ztz'] = alpha * self.constants['ztz'] + ztz
-        self.constants['ztX'] = alpha * self.constants['ztX'] + ztX
-
-        # Make sure the activation is not all 0
-        z_nnz = np.sum(z_hat != 0, axis=(0, 2))
-
-        if self.verbose > 5:
-            print("[{}] sparsity: {:.3e}".format(
-                self.name, z_nnz.sum() / z_hat.size))
-
-        if np.all(z_nnz == 0):
-            # No need to update the dictionary as this batch has all its
-            # activation to 0.
-            import warnings
-            warnings.warn("Regularization parameter `reg` is too large and all"
-                          " the activations are zero. The atoms has not been "
-                          "updated.", UserWarning)
-            return z_hat
-
-        d_kwargs = dict(verbose=self.verbose, eps=1e-8)
-        d_kwargs.update(self.solver_d_kwargs)
-        if self.rank1:
-            self._D_hat = update_uv(
-                X, z_hat, uv_hat0=self._D_hat, constants=self.constants,
-                solver_d=self.solver_d, uv_constraint=self.uv_constraint,
-                loss=self.loss, loss_params=self.loss_params,
-                window=self.window, **d_kwargs
-            )
-        else:
-            self._D_hat = update_d(
-                X, z_hat, D_hat0=self._D_hat, constants=self.constants,
-                solver_d=self.solver_d, window=self.window,
-                loss=self.loss, loss_params=self.loss_params, **d_kwargs
-            )
-
-        return z_hat
+            z_encoder.compute_z()
+
+            alpha = self.algorithm_params['alpha']
+            self.constants['ztz'] = alpha * \
+                self.constants['ztz'] + z_encoder.ztz
+            z_encoder.ztz = self.constants['ztz']
+            self.constants['ztX'] = alpha * \
+                self.constants['ztX'] + z_encoder.ztX
+            z_encoder.ztX = self.constants['ztX']
```

Comment:
I think it would be more efficient to only inline operations instead of allocating 3 arrays?
```suggestion
            self.constants['ztz'] *= alpha
            self.constants['ztz'] += z_encoder.ztz
            z_encoder.ztz = self.constants['ztz']
            self.constants['ztX'] *= alpha
            self.constants['ztX'] += z_encoder.ztX
            z_encoder.ztX = self.constants['ztX']
```

Also, I am not sure why we should